File: backend/app/repositories/audit_log_repository.py
"""Query mechanics for `audit_log` only — no business logic. A tiny,
separate file rather than bundled into whichever service first needed
it (ConfigurationService): audit_log is a cross-cutting, append-only
table every future service will eventually write to, not something
that belongs conceptually to configuration. arcnave_app has SELECT/
INSERT only on this table (no UPDATE/DELETE, by design — see the
Module 0 migration) — an audit trail the app role can rewrite or erase
isn't an audit trail, so create_audit_log_entry is the only write this
file offers.
"""
import json

from sqlalchemy import text
from sqlalchemy.orm import Session


def create_audit_log_entry(
    db: Session,
    *,
    college_id: str,
    user_id: str,
    action: str,
    entity: str,
    entity_id: str,
    metadata: dict,
) -> None:
    db.execute(
        text(
            "INSERT INTO audit_log (college_id, user_id, action, entity, entity_id, metadata) "
            "VALUES (:college_id, :user_id, :action, :entity, :entity_id, :metadata)"
        ),
        {
            "college_id": college_id,
            "user_id": user_id,
            "action": action,
            "entity": entity,
            "entity_id": entity_id,
            "metadata": json.dumps(metadata),
        },
    )

File: backend/src/repositories/auditLogRepository.js
'use strict';

// Query mechanics for `audit_log` only — no business logic. A tiny,
// separate file rather than bundled into whichever service first
// needed it (configurationService.js): audit_log is a cross-cutting,
// append-only table every future service will eventually write to,
// not something that belongs conceptually to configuration.
// arcnave_app has SELECT/INSERT only on this table (no UPDATE/DELETE,
// by design — see the ported migration) — an audit trail the app
// role can rewrite or erase isn't an audit trail, so
// createAuditLogEntry is the only write this file offers.

const { getRequestContext } = require('../logging/context');

// Identity-Architecture.md Audit Identity: the Acting Position Account
// and Position, when the action happened in a position context.
// Deliberately NOT threaded as an explicit parameter through every one
// of this repository's ~100 call sites — middleware/identity.js
// mutates the same AsyncLocalStorage request context logger.js already
// reads req.collegeId/requestId from, so this defaults from ambient
// context exactly the way those already do, and every existing and
// future call site gets a correct value for free. A caller that
// already resolved capabilities itself (or is running with no
// position context on purpose) can still pass positionAccountId/
// positionId explicitly — an explicit `null` is respected, never
// overridden by the ambient default; only an *omitted* key falls back
// to it.
function ambientPosition() {
  const context = getRequestContext();
  const capabilities = context ? context.capabilities : null;
  if (!capabilities) {
    return { positionAccountId: null, positionId: null };
  }

  // Phase 2: a Position Account session's capabilities (identityService.
  // resolveCapabilitiesForPosition) carry positionAccountId/positionId
  // directly — there is no `.positions` array, unlike a personal-login
  // session's resolveCapabilities shape below, since exactly one
  // position is ever in scope for that session.
  if (capabilities.positionAccountId !== undefined) {
    return { positionAccountId: capabilities.positionAccountId, positionId: capabilities.positionId };
  }

  const positions = capabilities.positions;
  if (!positions || positions.length === 0) {
    return { positionAccountId: null, positionId: null };
  }
  // positionResolver orders by level ASC — the same "lowest level
  // number wins" tie-break identityService.deriveEffectiveRole already
  // applies for effectiveRole, kept consistent here.
  const [primary] = positions;
  return { positionAccountId: primary.positionAccountId, positionId: primary.positionId };
}

async function createAuditLogEntry(client, {
  collegeId, userId, action, entity, entityId, metadata, positionAccountId, positionId,
}) {
  const needsDefault = positionAccountId === undefined || positionId === undefined;
  const ambient = needsDefault ? ambientPosition() : null;
  const resolvedPositionAccountId = positionAccountId !== undefined ? positionAccountId : ambient.positionAccountId;
  const resolvedPositionId = positionId !== undefined ? positionId : ambient.positionId;

  await client.query(
    `INSERT INTO audit_log (college_id, user_id, action, entity, entity_id, metadata, position_account_id, position_id)
     VALUES ($1, $2, $3, $4, $5, $6, $7, $8)`,
    [collegeId, userId, action, entity, entityId, JSON.stringify(metadata), resolvedPositionAccountId, resolvedPositionId],
  );
}

module.exports = { createAuditLogEntry };
